go-controller/pkg/node/port_claim.go

```diff
@@ -73,38 +83,61 @@ func initPortClaimWatcher(recorder record.EventRecorder, wf *factory.WatchFactor
 }
 
 func addServicePortClaim(svc *kapi.Service) []error {
-	return handleService(svc, port.open)
+	return handleService(svc, portHandler.open)
 }
 
 func deleteServicePortClaim(svc *kapi.Service) []error {
-	return handleService(svc, port.close)
+	return handleService(svc, portHandler.close)
 }
 
 func handleService(svc *kapi.Service, handler handler) []error {
 	errors := []error{}
 	if !util.ServiceTypeHasNodePort(svc) && len(svc.Spec.ExternalIPs) == 0 {
 		return errors
 	}
+
 	for _, svcPort := range svc.Spec.Ports {
 		if util.ServiceTypeHasNodePort(svc) {
-			if err := handlePort(svc, svcPort.NodePort, svcPort.Protocol, handler); err != nil {
+			klog.V(5).Infof("Handle NodePort service %s port %d", svc.Name, svcPort.NodePort)
+			if err := handlePort(getDescription(svcPort.Name, svc, true), svc, "", svcPort.NodePort, svcPort.Protocol, handler); err != nil {
 				errors = append(errors, err)
 			}
 		}
-		if len(svc.Spec.ExternalIPs) > 0 {
-			if err := handlePort(svc, svcPort.Port, svcPort.Protocol, handler); err != nil {
+		for _, externalIP := range svc.Spec.ExternalIPs {
```

Comment:
This is wrong I think. We don't want to open multiple sockets to the same port, which is what will happen now if we have a list of external IPs. We only want to bind one socket to a local port as to block the port from being used by other processes on the node, as to guard against port conflict between host network pods / system services running on the node. I am actually very confused why upstream opens a port on external IP:port, here: https://github.com/kubernetes/utils/blob/master/net/port.go#L113 . It seems unnecessary.

Reply:
Let's assume that a host where a K8s node is running, has several local IPs, one of them is an "External IP" for a service A. Now a user wants to bind a service B, which can be a K8s service or an external process to the same port but to a different IP.

The current implementation prevents it, even worse, it creates a race condition. If the service B starts first, we will not be able to bind the service A guarded socket to all IPs , but this will not prevent the service from starting. On another hand, if the service A starts first and binds its guarded socket to all IPs , the service B will not be able to start. 

In addition, there is an inconsistency between iptables roles created with specific IP address restrictions (-d externalIP) and their guarded sockets which are bound to all IPs.

Reply:
Understood now. I've always found the external IP feature of Kubernetes a bit contrived and difficult to wrap my head around, but what you say makes sense.